src/ApiService/ApiService/Log.cs

```diff
@@ -0,0 +1,172 @@
+﻿using System;
+using System.Collections.Generic;
+using Microsoft.ApplicationInsights;
+using Microsoft.ApplicationInsights.Extensibility;
+using Microsoft.ApplicationInsights.DataContracts;
+
+using System.Diagnostics;
+
+namespace Microsoft.OneFuzz.Service;
+
+public interface ILog {
+    void Log(Guid correlationId, String message, SeverityLevel level, IDictionary<string, string> tags, string? caller);
+    void LogEvent(Guid correlationId, String evt, IDictionary<string, string> tags, IDictionary<string, double>? metrics, string? caller);
+    void LogException(Guid correlationId, Exception ex, IDictionary<string, string> tags, IDictionary<string, double>? metrics, string? caller);
+    void Flush();
+}
+
+class AppInsights : ILog {
+    private TelemetryClient telemetryClient =
+            new TelemetryClient(
+                new TelemetryConfiguration(EnvironmentVariables.AppInsights.InstrumentationKey));
```

Comment:
can we use a contructor paramter instead of reading from an environment variable here. This makes the code easier to test

Reply:
Microsoft.OneFuzz.Service.EnvironmentVariables.LogDestinations <- [| Microsoft.OneFuzz.Service.LogDestination.Console |]

let log = Microsoft.OneFuzz.Service.Log(Guid.NewGuid())
log.Tags.Add("Tag1", "V1")
log.Tags.Add("Tag2", "V2")

log.Info("blah")

Reply:
just pre-set destination before calling your test.

Passing through constructor - means that every time you create logger object in azure function that is configured to do APpInsights, you have to pass AppInsights in the constructor.


This Log object meant to be an execution flow context type of thing, that tracks correlation ID and tags

Reply:
> Microsoft.OneFuzz.Service.EnvironmentVariables.LogDestinations <- [| Microsoft.OneFuzz.Service.LogDestination.Console |]

This make it hard to have two different instances of loggers 


> Passing through constructor - means that every time you create logger object in azure function that is configured to do APpInsights, you have to pass AppInsights in the constructor.

No you can create the logger once when you start the application then pass it around. That is the whole point of dependency injection. If you look at `QueueNodeHearbeat` you will see that it received an ilogger

Reply:
1) Why do we need 2 loggers writing to different destinations ? 

2) we need to track correlation ID across the flow of calls, and tags. 
So we know in which order which function got called.